File: go/internal/chunk/chunk.go
package chunk

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"

	"minidss/internal/proto"
)

type Chunker struct {
	BlockSize int
}

func New(blockSize int) *Chunker {
	if blockSize <= 0 {
		blockSize = proto.DefaultBlockSize
	}
	return &Chunker{BlockSize: blockSize}
}

// Manifest reads the file at path and produces a Manifest with per-block
// SHA-256 and a whole-file SHA-256.
func (c *Chunker) Manifest(path, name string) (*proto.Manifest, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	info, err := f.Stat()
	if err != nil {
		return nil, err
	}

	fileH := sha256.New()
	buf := make([]byte, c.BlockSize)
	var blocks []proto.BlockInfo
	idx := 0
	for {
		n, err := io.ReadFull(f, buf)
		if n > 0 {
			bh := sha256.New()
			bh.Write(buf[:n])
			fileH.Write(buf[:n])
			blocks = append(blocks, proto.BlockInfo{
				Index:  idx,
				Size:   n,
				SHA256: hex.EncodeToString(bh.Sum(nil)),
			})
			idx++
		}
		if err == io.EOF || err == io.ErrUnexpectedEOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read: %w", err)
		}
	}

	return &proto.Manifest{
		Name:      name,
		Size:      info.Size(),
		SHA256:    hex.EncodeToString(fileH.Sum(nil)),
		BlockSize: c.BlockSize,
		Blocks:    blocks,
	}, nil
}

func HashBytes(b []byte) string {
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:])
}

File: ts/src/chunk.ts
// File chunking + hashing. Mirrors go/internal/chunk.

import { createHash } from "node:crypto";
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import { DEFAULT_BLOCK_SIZE, type BlockInfo, type Manifest } from "./proto.js";

export function sha256Hex(buf: Uint8Array): string {
  return createHash("sha256").update(buf).digest("hex");
}

/**
 * Build a Manifest for the file at `path`, splitting into fixed-size blocks.
 * Computes a per-block SHA-256 and a whole-file SHA-256 in a single pass.
 */
export async function buildManifest(
  path: string,
  name: string,
  blockSize: number = DEFAULT_BLOCK_SIZE,
): Promise<Manifest> {
  if (blockSize <= 0) blockSize = DEFAULT_BLOCK_SIZE;

  const info = await stat(path);
  const fileHash = createHash("sha256");
  const blocks: BlockInfo[] = [];

  let index = 0;
  let carry: Buffer = Buffer.alloc(0);

  const emit = (chunk: Buffer): void => {
    fileHash.update(chunk);
    blocks.push({
      index: index++,
      size: chunk.length,
      sha256: createHash("sha256").update(chunk).digest("hex"),
    });
  };

  await new Promise<void>((resolve, reject) => {
    // Read in modest pieces; re-slice into exact blockSize blocks so the
    // manifest is independent of the OS read granularity.
    const stream = createReadStream(path, { highWaterMark: blockSize });
    stream.on("data", (data: string | Buffer) => {
      const piece = typeof data === "string" ? Buffer.from(data) : data;
      carry = carry.length === 0 ? piece : Buffer.concat([carry, piece]);
      while (carry.length >= blockSize) {
        emit(carry.subarray(0, blockSize));
        carry = carry.subarray(blockSize);
      }
    });
    stream.on("end", () => {
      if (carry.length > 0) emit(carry);
      resolve();
    });
    stream.on("error", reject);
  });

  return {
    name,
    size: info.size,
    sha256: fileHash.digest("hex"),
    block_size: blockSize,
    blocks,
  };
}
